compiler/hash-ir/src/ty.rs

```diff
@@ -0,0 +1,202 @@
+//! Simplified Hash type hierarchy. The [IrTy] is used a simplified
+//! variant of the available Hash [Term]s that are defined in the
+//! `hash-types` crate. This data structure is used to remove unnecessary
+//! complexity from types that are required for IR generation and
+//! analysis.
+
+use std::fmt;
+
+use bitflags::bitflags;
+use hash_source::{
+    constant::{FloatTy, SIntTy, UIntTy},
+    identifier::Identifier,
+};
+use hash_utils::{
+    new_sequence_store_key, new_store_key,
+    store::{CloneStore, DefaultSequenceStore, DefaultStore, SequenceStore, Store},
+};
+use index_vec::IndexVec;
+
+use crate::write::{ForFormatting, WriteTyIr};
+
+/// Simplified type structure used by the IR and other stages to reason about
+/// Hash programs once types have been erased and simplified.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum IrTy {
+    /// A primitive signed integer type, e.g. `i32`.
+    Int(SIntTy),
+
+    /// A primitive unsigned integer type, e.g. `u32`.
+    UInt(UIntTy),
+
+    /// A primitive floating point type, e.g. `f32`.
+    Float(FloatTy),
+
+    /// String type.
+    Str,
+
+    /// Boolean type.
+    Bool,
+
+    /// Character type.
+    Char,
+
+    /// The never type
+    Never,
+
+    /// A reference type, referring to a another type, e.g. `&T`
+    Ref(IrTyId),
+    /// A tuple type,
+    Tuple(IrTyListId),
+
+    /// An abstract data structure type, i.e a `struct` or `enum`, or any
+    /// other kind of type.
+    Adt(AdtId),
+
+    /// The first item is the interned parameter types to the function, and the
+    /// second item is the return type of the function. If the function has no
+    /// explicit return type, this will always be inferred at this stage.
+    Fn(IrTyListId, IrTyId),
+}
+
+index_vec::define_index_type! {
+    /// Index for [VariantIdx] stores within generated [Body]s.
+    pub struct VariantIdx = u32;
+
+    MAX_INDEX = i32::max_value() as usize;
+    DISABLE_MAX_INDEX_CHECK = cfg!(not(debug_assertions));
+
+    DEBUG_FORMAT = "variant#{}";
+}
+
+#[derive(Clone)]
+pub struct AdtData {
+    /// The name of the ADT
+    pub name: Identifier,
+
+    /// All of the variants that are defined for this variant.
+    pub variants: IndexVec<VariantIdx, AdtVariant>,
+
+    /// Flags which denote additional information about this specific
+    /// data structure.
+    pub flags: AdtFlags,
+
+    /// Options that are regarding the representation of the ADT
+    /// in memory.
+    pub representation: AdtRepresentation,
+}
+
+bitflags! {
+    /// Defines the flags that denote which kind of ADT is being represented with
+    /// [AdtData].
+    ///
+    /// @@Todo: do we need add specific flags for `data` variant??
+    pub struct AdtFlags: u32 {
+        /// The ADT is a enum.
+        const ENUM = 0b00000001;
+
+        /// The ADT is a struct.
+        const STRUCT = 0b00000010;
```

Comment:
These are mutually exclusive, does it really make sense to use bitflags for it?

Reply:
Also this information is not retained by the typechecker at all anyway, only perhaps for error message purposes, so it shouldn't be relevant at this stage---both structs and enums desugar to `datatype` definitions.